Get offline mode launching with snap

superm1 opened the pull request:
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

superm1 commented:
@Sean-StarLabs can you please see if this one works for you in snap form?

Sean-StarLabs commented:
[log.txt](https://github.com/hughsie/fwupd/files/3135907/log.txt)
Testing on my laptop doesn't work, I can see:
```
May 01 21:17:15 sean-LabTop systemd[1]: fwupd-offline-update.service: Succeeded.
```
But EC is untouched

superm1 commented:
Can you share `journalctl` output for `fwupd-offline-update.service`?

Sean-StarLabs commented:
Sure:
```
May 01 21:13:40 sean-LabTop systemd[1]: Starting Updates device firmware whilst offline...
May 01 21:13:40 sean-LabTop systemd[1]: fwupd-offline-update.service: Succeeded.
May 01 21:13:40 sean-LabTop systemd[1]: Started Updates device firmware whilst offline.
-- Reboot --
May 01 21:17:15 sean-LabTop systemd[1]: Starting Updates device firmware whilst offline...
May 01 21:17:15 sean-LabTop systemd[1]: fwupd-offline-update.service: Succeeded.
May 01 21:17:15 sean-LabTop systemd[1]: Started Updates device firmware whilst offline.
```

superm1 commented:
Hmm.  I was expecting to see some error there.  I wonder if the daemon isn't seeing it for some reason at bootup?

1. Modify `/root/snap/fwupd/current/etc/fwupd/daemon.conf` to put `VerboseDomains` to `*`
2. Re-run the offline update
2. Reboot the system.
4. Check the journal for `snap.fwupd.fwupd` at the same time that `fwupd-offline-update.service` ran

Sean-StarLabs commented:
[fwupd-offline-update.service.log.txt](https://github.com/hughsie/fwupd/files/3135982/fwupd-offline-update.service.log.txt)
[snap.fwupd.fwupd.log.txt](https://github.com/hughsie/fwupd/files/3135983/snap.fwupd.fwupd.log.txt)

Not much info there?

superm1 commented:
It seems that there is not a corresponding entry for the two times fwupd ran with `Install` argument (`21:47`) and (`21:48`) in `fwupd-offline-update`.

I also don't see any message that is `schedule %s to be installed to %s on next boot` which I expected.  Or really anything about scheduling..

Did the frontend show any error while trying to do this?

Sean-StarLabs commented:
No errors, just:
```
1f664ad149f2151fb750b52751dbf9ba045a7846 is already scheduled to be updated
```

superm1 commented:
Oh, that's a pretty significant error :)

What does the symlink in `/system-update` point to?  I'm guessing it's not valid (due to earlier development fixes).

If it's invalid, remove it and try again.

Sean-StarLabs commented:
Do you mean in:
```
/usr/lib/systemd/system/system-update.target.wants
```

```
lrwxrwxrwx 1 root root 31 Apr 29 12:15 fwupd-offline-update.service -> ../fwupd-offline-update.service
lrwxrwxrwx 1 root root 36 Mar  2 21:02 packagekit-offline-update.service -> ../packagekit-offline-update.service
```

superm1 commented:
No I mean literally in the root of the filesystem:
``
/system-update
```

That symlink is the magic thing that triggers updates to happen.  If it's invalid it will lead to this behavior described.  And I expect it to be invalid for you because it was hardcoded to `/var/lib/fwupd` until recent commits I made.

Sean-StarLabs commented:
Looked there but nothing that matches:
```
lrwxrwxrwx   1 root root          7 Apr  4 10:53 bin -> usr/bin
drwxr-xr-x   4 root root       4096 Apr 17 09:08 boot
drwxrwxr-x   2 root root       4096 Apr  4 10:54 cdrom
-rw-r--r--   1 root root         57 Apr 29 12:50 changelog
drwxr-xr-x  19 root root       4700 May  1 22:04 dev
drwxr-xr-x 152 root root      12288 May  1 11:33 etc
-rw-r--r--   1 root root        146 Apr 29 12:52 files.json
drwxr-xr-x   3 root root       4096 Apr  4 10:54 home
lrwxrwxrwx   1 root root         32 Apr 17 08:20 initrd.img -> boot/initrd.img-5.0.0-13-generic
lrwxrwxrwx   1 root root         32 Apr 17 08:20 initrd.img.old -> boot/initrd.img-5.0.0-11-generic
lrwxrwxrwx   1 root root          7 Apr  4 10:53 lib -> usr/lib
lrwxrwxrwx   1 root root          9 Apr  4 10:53 lib32 -> usr/lib32
lrwxrwxrwx   1 root root          9 Apr  4 10:53 lib64 -> usr/lib64
lrwxrwxrwx   1 root root         10 Apr  4 10:53 libx32 -> usr/libx32
drwx------   2 root root      16384 Apr  4 10:52 lost+found
drwxr-xr-x   3 root root       4096 Apr  4 10:59 media
drwxr-xr-x   2 root root       4096 Mar 26 22:22 mnt
drwxr-xr-x   4 root root       4096 Apr 29 09:38 opt
dr-xr-xr-x 300 root root          0 May  1 22:03 proc
drwx------   5 root root       4096 Apr 18 12:53 root
drwxr-xr-x  34 root root       1000 May  1 22:04 run
lrwxrwxrwx   1 root root          8 Apr  4 10:53 sbin -> usr/sbin
drwxr-xr-x  18 root root       4096 May  1 21:16 snap
drwxr-xr-x   2 root root       4096 Mar 26 22:22 srv
-rw-------   1 root root 2147483648 Apr  4 10:52 swapfile
dr-xr-xr-x  13 root root          0 May  1 22:03 sys
drwxrwxrwt  22 root root       4096 May  1 22:14 tmp
drwxr-xr-x  14 root root       4096 Mar 26 22:23 usr
drwxr-xr-x  14 root root       4096 Mar 26 22:30 var
lrwxrwxrwx   1 root root         29 Apr 17 08:20 vmlinuz -> boot/vmlinuz-5.0.0-13-generic
lrwxrwxrwx   1 root root         29 Apr 17 08:20 vmlinuz.old -> boot/vmlinuz-5.0.0-11-generic
```

superm1 commented:
Well that doesn't make sense to me, one of the commits I made in this PR: 11374f7 specifically is supposed to fix that behavior.

Are you sure you built/pulled it right?

Big hammer solution is to just stop snap.fwupd.fwupd, remove `/root/snap/fwupd` and re-run it.

Sean-StarLabs commented:
Big hammer...
```
Apr 26 10:25:24 sean-LabTop systemd[1]: Starting Updates device firmware whilst offline...
Apr 26 10:25:27 sean-LabTop fwupdoffline[728]: No updates were applied
Apr 26 10:25:27 sean-LabTop systemd[1]: fwupd-offline-update.service: Main process exited, code=exited, status=1/FAILURE
Apr 26 10:25:27 sean-LabTop systemd[1]: fwupd-offline-update.service: Failed with result 'exit-code'.
Apr 26 10:25:27 sean-LabTop systemd[1]: Failed to start Updates device firmware whilst offline.

```
[log.txt](https://github.com/hughsie/fwupd/files/3136128/log.txt)

superm1 commented:
Can you check in `/etc/systemd/system/system-update.target.wants`?  I expect to see 3 things:
* snap.fwupd.fwupd.service
* snap-core-####.mount
* fwupd-offline-update.service

Sean-StarLabs commented:
I don't have the folder called /etc/systemd/system/system-update.target.wants?

Sean-StarLabs commented:
Just built the snap on my laptop and then tested on a "clean" laptop, the EC is updated on shutdown. It does seem to report as terminated but, its a very minor quirk for it working!
```
May 02 10:25:34 sean-LabTop systemd[1]: Stopped Service for snap application fwupd.fwupd.
May 02 10:25:34 sean-LabTop systemd[1]: fwupd-offline-update.service: Failed with result 'signal'.
May 02 10:25:34 sean-LabTop systemd[1]: fwupd-offline-update.service: Main process exited, code=killed, status=15/TERM
May 02 10:25:34 sean-LabTop systemd[1]: Stopping Service for snap application fwupd.fwupd...
May 02 10:25:34 sean-LabTop fwupd.fwupdoffline[806]: Updating SuperIO IT8987 from 03.06 to 3.07... Done!
May 02 10:25:19 sean-LabTop dbus-daemon[833]: [system] Successfully activated service 'org.freedesktop.fwupd'

```

superm1 commented:
> I don't have the folder called /etc/systemd/system/system-update.target.wants?

What OS was that on?

>May 02 10:25:34 sean-LabTop systemd[1]: Stopped Service for snap application fwupd.fwupd.
May 02 10:25:34 sean-LabTop systemd[1]: fwupd-offline-update.service: Failed with result 'signal'.
May 02 10:25:34 sean-LabTop systemd[1]: fwupd-offline-update.service: Main process exited, code=killed, status=15/TERM
May 02 10:25:34 sean-LabTop systemd[1]: Stopping Service for snap application fwupd.fwupd...

@hughsie is this some cleanup stuff that needs adjusting?  I would guess SIGTERM was sent by systemd.  Maybe instead of calling `fu_util_update_reboot` the `SuccessAction` should be set to `reboot`?

hughsie commented:
> the SuccessAction should be set to reboot

I guess that works; I can't remember why the script has to do it manually. Perhaps before SuccessAction existed?

Sean-StarLabs commented:
> What OS was that on?

Zorin OS 15

superm1 commented:
> Zorin OS 15

OK I modified to make sure that there is a test for this directory before starting.

> I guess that works; I can't remember why the script has to do it manually. Perhaps before SuccessAction existed?

Alright, added a commit to do this with SuccessAction instead.

superm1 commented:
> Alright, added a commit to do this with SuccessAction instead.

Before we merge this I was thinking about it, and I have a thought why `SuccessAction` is wrong here.  In the case that system-update is running, don't all the things in `system-update.wants` run?  So that means that if there is a `packagekit-offline-update.service` in that directory it tries to run, then `fwupd-offline-update.service` tries to run etc.  If `/system-update` is set to packagekit's directory then `fwupd-offline-update` will still return `EXIT_SUCCESS` which is wrong in that case..

hughsie commented:
> If /system-update is set to packagekit's directory then fwupd-offline-update will still return EXIT_SUCCESS which is wrong in that case..

Urgh, that's exactly why we don't use SuccessAction -- sorry for not remembering!

superm1 commented:
> Urgh, that's exactly why we don't use SuccessAction -- sorry for not remembering!

I pushed an alternative.  Can you review it?

@Sean-StarLabs 

Can you give one more check with what's been force pushed to the branch?  I believe this should fix the `fwupd-offline-update` getting `SIGTERM`.

Sean-StarLabs commented:
It still triggers a second boot and then the grub prompt, different result though:

```
May 03 15:56:59 sean-LabTop systemd[1]: fwupd-offline-update.service: Failed with result 'exit-code'.
May 03 15:56:59 sean-LabTop systemd[1]: fwupd-offline-update.service: Main process exited, code=exited, status=1/FAILURE
```

It does work though!

[log.txt](https://github.com/hughsie/fwupd/files/3142423/log.txt)

superm1 commented:
OK one more try, it looks like `FailureActionExitStatus` needs a code (probably 0).
Multiple boots is definitely expected though.

Sean-StarLabs commented:
Sorry, didn't mean the second boot was unexpected, just that it occurs after the second reboot. The first reboot is "normal".

superm1 commented:
> Sorry, didn't mean the second boot was unexpected, just that it occurs after the second reboot. 

I think that might be due to an Ubuntu issue around `system-update`.  On Ubuntu there is a `grub-initrd-fallback.service` which will be called at the end of bootup to indicate it was a successful boot.  That suppresses the GRUB menu on next boot.

This isn't called in `system-update.target`.  I've filed this bug with Ubuntu about it.
https://bugs.launchpad.net/ubuntu/+source/grub2/+bug/1827532

I believe you can experiment with that solution by creating a symlink to that unit file in the directory `/lib/systemd/system/system-update.target.wants`.

superm1 commented:
@Sean-StarLabs Can you please try one more time with the `FailureActionExitStatus ` change I made?  I think this should be good now (other than the GRUB menu thing).

If things look good , then can you please try to make that symlink I mentioned? If it works, you should add a comment to the Ubuntu bug I referenced above.

@hughsie if you can please re-review, I believe this is in good shape now.

hughsie commented on src/fu-plugin.c:
```diff
@@ -861,17 +863,15 @@ fu_plugin_runner_offline_invalidate (GError **error)
 
 static gboolean
-fu_plugin_runner_offline_setup (GError **error)
+fu_plugin_runner_offline_setup (gchar *system_update, GError **error)
```
const?

hughsie commented on src/fu-plugin.c:
```diff
@@ -1359,6 +1399,17 @@ fu_plugin_runner_schedule_update (FuPlugin *self,
 	    fu_device_get_update_state (res_tmp) == FWUPD_UPDATE_STATE_PENDING) {
+		/* still pending but system-update was destroyed
+		 * (device unplugged at bootup?)
+		 **/
```
not gtk-doc end comment pls, i.e. on two line only.

hughsie commented on data/fwupd-offline-update.service.in:
```diff
@@ -11,3 +11,4 @@ Before=shutdown.target system-update.target
 Type=oneshot
 ExecStart=@libexecdir@/fwupd/fwupdoffline
 FailureAction=reboot
+FailureActionExitStatus=0
```
So we rewrite the exit status to 0 on failure? Not sure I understood the remove-reboot-exit-failure logic...

superm1 replied:
Let me try to explain it better.
1. In the situation that the offline target isn't intended to be fwupd, return `EXIT_SUCCESS`, but no reboot so that whatever the actual target is can run.
2. In the situation that fwupd is the offline target, need to reboot either way at the end.
3. If you call reboot before the application exits, then it gets `SIGTERM`


Do you think it would be clearer if it was inverted?
* So the calling fwupd-offline-update when something else is offline target, you return `EXIT_FAILURE`.
* In all other cases return `EXIT_SUCCESS`
* Adjust the systemd unit to reboot on `SuccessAction` instead of `FailureAction`

hughsie replied:
tbh, it all seems very complicated. I wonder if we should just not use the offline mode at all -- it's even causing problems on OSTree systems as `/` is immutable. What about if fwupdoffline was just run it with `WantedBy=final.target` and we did the upgrade on shutdown rather than with two new reboots? Two birds one stone.

The super important bit for me is the forced restart of the keyboard controller after flashing (as the fans might not work until we do!) but I think I can reset any EC parameters set before flashing. Before I test if this is even possible, I guess this is an order of magnitude easier to implement for the snap?

superm1 replied:
> What about if fwupdoffline was just run it with WantedBy=final.target and we did the upgrade on shutdown rather than with two new reboots? Two birds one stone.

If it was moved to shutdown then I sort of think we may as well just combine the activate and offline job.  In many ways they're similar, calling into a device's vfunc to do some action to the device.

I do like that for offline updates it was essentially `ln -s /system-update $SOMEWHERE` and it only runs that target during that situation.  I think it can be adopted still.  So how about something like this in the systemd job that would run on shutdown:

`ConditionPathExists=@localstatedir@/lib/fwupd/offline-pending`

And the the first thing the daemon does next time it's started is removes that file?

>  Before I test if this is even possible, I guess this is an order of magnitude easier to implement for the snap?

Yes, I do think so.

hughsie replied:
> we may as well just combine the activate and offline job

Not quite; the shutdown job would be run with /usr still mounted and the activation would be run with /usr unmounted.

> ConditionPathExists=@localstatedir@/lib/fwupd/offline-pending

Yes, I think that makes sense. Give me an hour or so.

hughsie replied:
Ideas please: How could the daemon "know" it's running in offline mode, i.e. before `shutdown.target`? We can't do `{GetUnit}.Get(ActiveState)` as the unit is not (yet) on the bus when the D-Bus method is called. I really want to avoid the client saying to the daemon "I'm offline now, TRUST ME" and work it out in the daemon.

hughsie replied:
Actually, just a bug, no race.

hughsie replied:
I've pushed a branch wip/hughsie/offline-at-shutdown with some ideas. Zero testing, but I thought it might be a good way to start thinking about how to do it properly at shutdown. I think you two understand the systemd unit more than me.

Sean-StarLabs commented:
```May 07 08:56:59 sean-LabTop systemd[1]: fwupd-offline-update.service: Failed with result 'exit-code'.
May 07 08:56:59 sean-LabTop systemd[1]: fwupd-offline-update.service: Main process exited, code=exited, status=1/FAILURE
May 07 08:56:59 sean-LabTop fwupd.fwupdoffline[759]: Downgrading SuperIO IT8987 from 03.07 to 3.06... Done!
May 07 08:56:43 sean-LabTop dbus-daemon[760]: [system] Successfully activated service 'org.freedesktop.fwupd'
```
Still shows 'Failure'. That is what you're trying to fix right?

[log.txt](https://github.com/hughsie/fwupd/files/3151423/log.txt)

superm1 commented:
Maybe use fwupdtool and set it up in the environment variables for the systemd unit?

On May 7, 2019 09:15, Richard Hughes <notifications@github.com> wrote:

[EXTERNAL EMAIL]

@hughsie commented on this pull request.

________________________________

In data/fwupd-offline-update.service.in<https://github.com/hughsie/fwupd/pull/1180#discussion_r281651286>:

> @@ -11,3 +11,4 @@ Before=shutdown.target system-update.target
 Type=oneshot
 ExecStart=@libexecdir@/fwupd/fwupdoffline
 FailureAction=reboot
+FailureActionExitStatus=0


Ideas please: How could the daemon "know" it's running in offline mode, i.e. before shutdown.target? We can't do {GetUnit}.Get(ActiveState) as the unit is not (yet) on the bus when the D-Bus method is called. I really want to avoid the client saying to the daemon "I'm offline now, TRUST ME" and work it out in the daemon.

-
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/hughsie/fwupd/pull/1180#discussion_r281651286>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AADYXPVR2RMNNJTQQJWBMNTPUGFGXANCNFSM4HJVMZJQ>.
